### PR TITLE
Add `matches!` macro.

### PR DESCRIPTION
**Pull request by utkarshkukreti**
Code lifted from https://github.com/SimonSapin/rust-std-candidates, I made some minor changes and added docs and examples.

I hope you're fine with that, @SimonSapin?

### CODE REVIEW

File: src/matches.rs [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+/// # fn main() {
+/// assert!(matches!(2, 1 | 2 | 3));
+/// assert!(matches!('x', 'a' ... 'z'));
+/// assert!(!matches!(Some(1), None));
+/// assert!(matches!(Some(42), Some(n) if n == 42));
+/// # }
+/// ```
+#[macro_export]
+macro_rules! matches {
+    ($expr:expr, $($pat:tt)+) => {
```
Comment (reem) [inline src/matches.rs:19]: shouldn't these be `pat`s not `tt`s?

**Comment by SimonSapin** [comment]
> I hope you're fine with that, @SimonSapin?

Sure.

File: src/matches.rs (hunk 1 above)
Comment (SimonSapin) [inline src/matches.rs:19]: It should be `$( $patter: pat )|+`, but that’s forbidden by https://github.com/rust-lang/rfcs/blob/master/text/0550-macro-future-proofing.md . This is why `_tt_as_expr_hack` exists.

Comment (reem) [inline src/matches.rs:19]: I see. Interesting workaround.

**Comment by reem** [comment]
Alright, LGTM. I have definitely wanted this macro in my arsenal.

**Pull request merged**